eva/executor/groupby_executor.py

```diff
@@ -0,0 +1,63 @@
+# coding=utf-8
+# Copyright 2018-2022 EVA
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import re
+from typing import Iterator
+
+import pandas as pd
+
+from eva.executor.abstract_executor import AbstractExecutor
+from eva.models.storage.batch import Batch
+from eva.planner.groupby_plan import GroupByPlan
+
+
+class GroupByExecutor(AbstractExecutor):
+    """
+    Samples uniformly from the rows.
+
+    Arguments:
+        node (AbstractPlan): The Sample Plan
+
+    """
+
+    def __init__(self, node: GroupByPlan):
+        super().__init__(node)
+        # match the pattern of group by clause (e.g., 16f or 8s)
+        pattern = re.search(r"^\d+[fs]$", node.groupby_clause.value)
+        # if valid pattern
+        if pattern:
+            match_string = pattern.group(0)
+            if match_string[-1] == "f":
+                self._segment_length = int(match_string[:-1])
+                # TODO ACTION condition on segment length?
+            else:
+                err_msg = "Only grouping by frames (f) is supported"
+                raise ValueError(err_msg)
+        else:
+            err_msg = "Incorrect GROUP BY pattern: {}".format(node.groupby_clause.value)
+            raise ValueError(err_msg)
+
+    def validate(self):
+        pass
+
+    def exec(self) -> Iterator[Batch]:
+        child_executor = self.children[0]
+
+        buffer = Batch(pd.DataFrame())
+        for batch in child_executor.exec():
+            new_batch = buffer + batch
+            while len(new_batch) >= self._segment_length:
+                yield new_batch[: self._segment_length]
+                new_batch = new_batch[self._segment_length :]
+            buffer = new_batch
```

Comment:
Missed yielding the buffer outside the loop

Reply:
Do you mean for the remainder segment? We discard the final segment (of size < segment_length) because it cannot be processed by the UDF. The assumption is that all the segments are of equal size, and `GROUP BY '8f'` is a hard constraint (standard practice in action localization).

Reply:
Done.